### PR TITLE
Crypto agility updates

### LINKED ISSUE #548: cryptoAgility: make AccountID a union

### LINKED ISSUE #561: cryptoAgility: get rid of uint512
Instead move to things like
struct StellarSig {
           opaque bytes[64];
};

While doing this, ensure that we use proper tagging in the various sign/verify pairs so that we can't mix up different message types (like a signed SCP message and a signed tx, signed with the same key)

### LINKED ISSUE #563: typo: inflationPayout should be InflationPayout

### PR DESCRIPTION
**Pull request by MonsieurNicolas**
Resolves #548 : 
extensibility for crypto scheme used for signing. moved to a union for public key, variable length signatures
Proper domain partitioning for digital signatures

Also resolves #561 (get rid of uint512)

Also
 * fixed typo in .x file - resolves #563 

@nullstyle : this will break Horizon in a way less trivial than usual. Pay particular attention to the "domain separation for digital signature" commit that changes the way we sign transactions with an extra prefix when we compute the hash of the transaction (it's like an implicit uint32 at the beginning of the transaction).

### CODE REVIEW

**Comment by nullstyle** [comment]
@MonsieurNicolas could explain a little bit about why we're adding the prefix to the "content to be hashed"?

Did you consider modeling this prefix as an actual union?

```
union SignatureBase switch (EnvelopeType type) {
  case ENVELOPE_TYPE_SCP:
      SCPStatement statement;
  case ENVELOPE_TYPE_TX:
      Transaction tx;
};
```

**Comment by MonsieurNicolas** [comment]
It's possible to do it like this but then we're wasting 4 additional bytes per signature

**Comment by nullstyle** [comment]
By my count the xdr-encoded form of my proposed union will result in the exact same byte stream as the c++ `xdr::xdr_to_opaque(ENVELOPE_TYPE_TX, mEnvelope.tx)`.  

Perhaps you misunderstood me:  I'm not advocating that we take on a prefix uint32 to every signature, but rather I am advocating that we push another data structure that is important for the protocol into the .x files.

**Comment by nullstyle** [comment]
To further clarify: `SignatureBase` is the thing that gets encoded to xdr, hashed with SHA256, then signed.

File: src/xdr/Stellar-ledger-entries.x [hunk 1]
```diff
@@ -145,4 +188,14 @@ case TRUSTLINE:
 case OFFER:
     OfferEntry offer;
 };
+
+// list of all envelope types used in the application
+// those are prefixes used when building signatures for
+// the respective envelopes
+enum EnvelopeTypes
```
Comment (nullstyle) [inline src/xdr/Stellar-ledger-entries.x]: Probably want to singularize this, so it's consistent with the other enums.

File: src/xdr/Stellar-types.x [hunk 2]
```diff
@@ -2,56 +2,34 @@
...
-typedef opaque AccountID[32];
-typedef opaque Thresholds[4];
-typedef string string32<32>;
 
-typedef uint64 SequenceNumber;
+typedef unsigned hyper uint64;
+typedef hyper int64;
 
-enum CurrencyType
+enum CryptoKeyTypes
```
Comment (nullstyle) [inline src/xdr/Stellar-types.x]: This should also probably be singularized.

**Comment by MonsieurNicolas** [comment]
Oh I see, the EnvelopeBase would be a helper to calculate the signature.

I am not sure it adds much clarity to the protocol. The person implementing it has to know that the signature is based on that one off struct and has to know which algorithm to follow for generating the signature.
Even before that change the scheme used for signing SCP messages and transactions is different (we have an extra SHA256 in the case of transactions).

**Comment by MonsieurNicolas** [comment]
good point on the enum names, I'll correct that

**Comment by nullstyle** [comment]
> I am not sure it adds much clarity to the protocol. The person implementing it has to know that the signature is based on that one off struct and has to know which algorithm to follow for generating the signature.

Yeah, that works for me.  I just figured I would put the idea out there for consideration if it hadn't already.

**Pull request merged**